clients/rust/tests/validator_stake_tokens.rs

```diff
@@ -522,3 +531,199 @@ async fn fail_stake_tokens_with_uninitialized_stake_account() {
+        Some(&context.payer.pubkey()),
+        &[&context.payer, &rewards_manager.owner],
+        context.last_blockhash,
+    );
+    let err = context
+        .banks_client
+        .process_transaction(tx)
+        .await
+        .unwrap_err();
+
+    // Then we expect an error.
+
+    assert_custom_error!(err, PaladinStakeProgramError::TotalStakeAmountExceedsLimit);
+}
+
+#[tokio::test]
+async fn fail_validator_stake_tokens_with_insufficient_staked_sol() {
+    let mut program_test = ProgramTest::new(
+        "paladin_stake_program",
+        paladin_stake_program_client::ID,
+        None,
+    );
+    program_test.add_program(
+        "paladin_rewards_program",
+        paladin_rewards_program_client::ID,
+        None,
+    );
+    let mut context = program_test.start_with_context().await;
+
+    // Given a config account and a validator stake with 2 SOL staked.
+
+    let config_manager = ConfigManager::new(&mut context).await;
+    let stake_manager = ValidatorStakeManager::new(&mut context, &config_manager.config).await;
+
+    let mut account = get_account!(context, stake_manager.stake);
+    let mut stake_account = ValidatorStake::from_bytes(account.data.as_ref()).unwrap();
+    // "manually" set the amount to 2 SOL
+    stake_account.total_staked_lamports_amount = 2 * 1_000_000_000;
+
+    account.data = stake_account.try_to_vec().unwrap();
+    context.set_account(&stake_manager.stake, &account.into());
+
+    // And we initialize the holder rewards accounts and mint 100 tokens.
+
+    let rewards_manager = RewardsManager::new(
+        &mut context,
+        &config_manager.mint,
+        &config_manager.mint_authority,
+    )
+    .await;
+
+    mint_to(
+        &mut context,
+        &config_manager.mint,
+        &config_manager.mint_authority,
+        &rewards_manager.token_account,
+        100,
+        9, // <- 9 decimals
+    )
+    .await
+    .unwrap();
+
+    // And we create the holder rewards account for the vault account.
+
+    create_holder_rewards(
+        &mut context,
+        &rewards_manager.pool,
+        &config_manager.mint,
+        &config_manager.vault,
+    )
+    .await;
+
+    // When we try to stake 50 tokens on a validator stake account with only 2 SOL staked.
+    //
+    // - raw amount to be staked: 50_000_000_000
+    // - current lamports (SOL) staked: 2_000_000_000
+    // - stake limit: 1.3 * 2_000_000_000 = 2_600_000_000
+    //
+    // Stake amount exceeds the limit: 50_000_000_000 > 2_600_000_000
+
+    let mut stake_ix = ValidatorStakeTokensBuilder::new()
+        .config(config_manager.config)
+        .validator_stake(stake_manager.stake)
+        .source_token_account(rewards_manager.token_account)
+        .token_account_authority(rewards_manager.owner.pubkey())
+        .mint(config_manager.mint)
+        .vault(config_manager.vault)
+        .token_program(spl_token_2022::ID)
+        .amount(50_000_000_000) // <- stake 50 tokens (raw value with 9 digits precision)
```

Comment:
nit: it's typically more interesting to do the exact limits in tests, so in this case, we'd want to do `2_600_000_001` for it to just exactly fail

Reply:
Nice, that is a really good point, thanks!